wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceAttributeValidator.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.workspace.server;
+
+import java.util.Map;
+import org.eclipse.che.api.core.ValidationException;
+
+/**
+ * Adds an ability to extends logic of workspace attributes validation.
+ *
+ * <p>It may needed since attributes may be used as configuration storage by different components.
+ * And that components may need to validate attributes.
+ *
+ * @author Sergii Leshchenko
+ */
+public interface WorkspaceAttributeValidator {
+
+  /**
+   * Validates if the specified workspace attributes does not contain invalid attributes according
+   * to implementors rules.
+   *
+   * @param attributes workspace attributes to validate
+   * @throws ValidationException when the specified workspace attributes is not valid
+   */
+  void validate(Map<String, String> attributes) throws ValidationException;
```

Comment:
> IMHO exception should handle exceptional state. The failed validation is not exceptional.


Agree and not =)

If Java would be Go we would `isValid, message`
Now we are able to go two ways:
1. Throw exception with a concrete message for a user.
2. Introduce some ValidationResult and two fields `isValid, message`.

I believe from a user perspective of view it's much better to use
`Hey, namespace attribute is longer than 63 characters` or `Hey, you're not permitted to use any namespace, only allowed - `sleshche-che`` instead of `Something is wrong with your workspace/attributes`

Would you go with 2nd approach? 
I'm not sure that with current architecture, it would brings a lot of profit, ValidationResult should be transformed to Exception maybe on another level like WorkspaceManager, WorkspaceService.
I'm not sure that we should invest time to rework like `createWorkspace` to one that returns ValidationResult...
Hope I described my opinion pretty clear - if no - feel free to ask details.

Reply:
yes, I've realized the same after I've submitted the review. 100% for result object. And yes, it might be transformed into the exception where appropriate. Preferably unchecked exception that will be catched where it is appropriate.

Reply:
@sparkoo to make sure I've understand your last comment. Do you want to ValidationResult instead of the exception?

Reply:
@skabashnyuk yes, I would prefer that.

Reply:
-1. Instantiating an object each time even in a hopefully much more common happy case IMHO outweighs the positives.

Java has exceptions for rich handling of exceptional cases. Boolean return type makes this look like C, result object makes this look like Go or Rust. Let's just not.

Reply:
Have the same POV as @sleshchenko and @metlos. I would like to keep ValidationException.

Reply:
> -1. Instantiating an object each time even in a hopefully much more common happy case IMHO outweighs the positives.

You don't have to instantiate success result. It could be static final instance.

> 
> Java has exceptions for rich handling of exceptional cases. Boolean return type makes this look like C, result object makes this look like Go or Rust. Let's just not.

yes, these newer languages learned from java mistakes.